libraries/AP_HAL_Linux/UARTDriver.cpp

```diff
@@ -284,44 +244,33 @@ void UARTDriver::set_blocking_writes(bool blocking)
  */
 bool UARTDriver::tx_pending()
 {
-    return !BUF_EMPTY(_writebuf);
+    return (_writebuf.available() > 0);
 }
 
 /*
   return the number of bytes available to be read
  */
 uint32_t UARTDriver::available()
 {
-    if (!_initialised) {
-        return 0;
-    }
-    uint16_t _tail;
-    return BUF_AVAILABLE(_readbuf);
+    return _readbuf.available();
 }
 
 /*
   how many bytes are available in the output buffer?
  */
 uint32_t UARTDriver::txspace()
 {
-    if (!_initialised) {
-        return 0;
-    }
-    uint16_t _head;
-    return BUF_SPACE(_writebuf);
+    return _writebuf.space();
 }
 
 int16_t UARTDriver::read()
 {
     uint8_t c;
-    if (!_initialised || _readbuf == NULL) {
-        return -1;
-    }
-    if (BUF_EMPTY(_readbuf)) {
+    if (!_initialised) {
         return -1;
     }
-    c = _readbuf[_readbuf_head];
-    BUF_ADVANCEHEAD(_readbuf, 1);
+    c = _readbuf.peek(0);
+    _readbuf.advance(1);
```

Comment:
> For a single byte, read will do a lot of operations to handle more complex case of buffer splits and so on.

And this is what I meant with:

> I understand that read will make more operations than necessary but I think that should be fixed in the ByteBuffer implementation and not working around here.

I believe that either the ByteBuffer read should detect when it is a simple byte and do the simple operations or it should have a read that only returns one byte. When I say workaround is that you are using functions that should be encapsulated by the ByteBuffer API.